### PR TITLE
Documentation infrastructure

### PR DESCRIPTION
**Pull request by amitabhverma**
On the fly generation of docs (User Guide, API References, Examples, etc.) using sphinx and readthedocs integration.

A preview can be viewed here: https://waveorder.readthedocs.io/en/latest/index.html

- The html pages are rendered from github markdown (.md) files and some are included in .rst files.
- The API References (Settings and Stokes) are rendered based on docstrings. 
- The Examples are pulled directly from source code files.

image

Further customization is possible as required. Some restructuring of the docs files within the repo (guides, examples, etc.) would benefit users.

### CODE REVIEW

**Review by talonchandler — COMMENTED** [review]
First look comments:
- Looking good! I think the major infrastructure pieces are present here which is great. 
- I would suggest changing the base branch from `main` to `gui_update` (within github is fine...no need to work through a rebase imo). As is this PR includes changes from both branchs, e.g. `jobs_mgmt.py` belongs to the other PR. This will help focus @srivarra's attention.  
- No need for a duplicate `poincare_swing.svg`, correct?

**Comment by talonchandler** [comment]
> I have not moved the existing .md files and docs/images but duplicated them under the sphinx_source dir. So yes, there is some duplication of files at the moment which needs to be addressed. I assumed there might be a need to keep a structure that works on github as well.

If they're still `.md` files, they're still viewable in github so I'm fine if they live in the `sphinx_source` directory. My main concern is that duplicated files will be difficult to maintain.

File: docs/sphinx_source/README.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+python waveorder/docs/examples/models/phase_thick_3d.py
+```
+
+(M1 users) `pytorch` has [incomplete GPU support](https://github.com/pytorch/pytorch/issues/77764),
+so please use `export PYTORCH_ENABLE_MPS_FALLBACK=1`
+to allow some operators to fallback to CPU if you plan to use GPU acceleration for polarization reconstruction.
+
+
+## Examples
+The [examples](https://github.com/mehta-lab/waveorder/tree/main/docs/examples) illustrate simulations and reconstruction for 2D QLIPP, 3D phase from brightfield, and 2D/3D PTI methods.
```
Comment (srivarra) [inline docs/sphinx_source/README.md:223]: Q - Clarification: Should this be directing users to the directory on Github or to the Examples page on the docs site?

File: docs/sphinx_source/api/settings.rst
Comment (srivarra) [inline docs/sphinx_source/api/settings.rst:1]: Thoughts on adding more documentation for the settings, like what each one does?

File: docs/sphinx_source/guide/QLIPP.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# waveorder
+
+`waveorder` is a collection of computational imaging methods. It currently provides QLIPP (quantitative label-free imaging with phase and polarization), phase from defocus, and fluorescence deconvolution.
+
+These are the kinds of data you can acquire with `waveorder` and QLIPP:
+
+https://user-images.githubusercontent.com/9554101/271128301-cc71da57-df6f-401b-a955-796750a96d88.mov
+
+https://user-images.githubusercontent.com/9554101/271128510-aa2180af-607f-4c0c-912c-c18dc4f29432.mp4
+
```
Comment (srivarra) [inline docs/sphinx_source/guide/QLIPP.md:10]: These should be embedded so a user doesn't have to download them to view a .mov / .mp4 / .gif etc...

File: docs/sphinx_source/guide/buyers-guide.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+- Read about the [OpenPolScope Hardware Kit](https://openpolscope.org/pages/OPS_Hardware.htm).
+- Complete the [OpenPolScope information request form](https://openpolscope.org/pages/Info_Request_Form.htm).
+
+### Buy individual components
+
+The components are listed in the order in which they process light. See the build video here to see how to assemble these components on your microscope.
+
+<video>
+   <source src="https://github.com/user-attachments/assets/a0a8bffb-bf81-4401-9ace-3b4955436b57" type="video/mp4">
+</video>
```
Comment (srivarra) [inline docs/sphinx_source/guide/buyers-guide.md:26]: Can we embed this like the video clips like on the README, it's too cool to not be viewable :(

https://github.com/user-attachments/assets/9c177f9b-b0ec-4034-9aab-f026f29669bc

File: docs/sphinx_source/guide/buyers-guide.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+| Circular polarizer | $350 | We suggest [a Thorlabs 532 nm, left-hand circular polarizer](https://www.thorlabs.com/thorproduct.cfm?partnumber=CP1L532).|
+| Liquid crystal compensator | $6,000 | Meadowlark optics LVR-42x52mm-VIS-ASSY or LVR-50x60mm-VIS-POL-ASSY. Although near-variants are listed in the [Meadowlowlark catalog](https://www.meadowlark.com/product/liquid-crystal-variable-retarder/), this is a custom part with two liquid crystals in a custom housing. [Contact Meadowlark](https://www.meadowlark.com/contact-us/) for a quote.|
+| Liquid crystal control electronics | $2,000 | [Meadowlark optics D5020-20V](https://www.meadowlark.com/product/liquid-crystal-digital-interface-controller/). Choose the high-voltage 20V version.
+| Liquid crystal adapter | $25-$500 | A 3D printed part that aligns the liquid crystal compensator in a microscope stand's illumination path. Check for your stand among the [OpenPolScope `.stl` files](https://github.com/amitabhverma/Microscope-LC-adapters/tree/main/stl_files) or [contact us](mailto:compmicro@czbiohub.org) for more options.|
```
Comment (srivarra) [inline docs/sphinx_source/guide/buyers-guide.md:34]: `$25-$500` ends up rendering `25-` as latex

File: .readthedocs.yaml
Comment (srivarra) [inline .readthedocs.yaml:1]: Comment: I'm surprised that commands are needed here to build the docs on RTD. huh, neat.
<details>
<summary>An example of what I'm describing</summary>

```yaml
# https://docs.readthedocs.io/en/stable/config-file/v2.html
version: 2
build:
  os: ubuntu-22.04
  tools:
    python: "3.12"
  commands:
    - asdf plugin add uv
    - asdf install uv latest
    - asdf global uv latest
    - uv sync --extra doc --frozen
    - uv run -m sphinx -T -b html -d docs/_build/doctrees -D language=en docs $READTHEDOCS_OUTPUT/html
sphinx:
  configuration: docs/conf.py
  # disable this for more lenient docs builds
  fail_on_warning: true
# python:
#   install:
#     - method: pip
#       path: .
#       extra_requirements:
#         - doc
```

</details>

File: docs/sphinx_source/guide/development-guide.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
+# `waveorder` development guide
+
+## Install `waveorder` for development
+
+1. Install [conda](https://github.com/conda-forge/miniforge) and create a virtual environment:
+
+    ```sh
+    conda create -y -n waveorder python=3.12
+    conda activate waveorder
+    ```
```
Comment (srivarra) [inline docs/sphinx_source/guide/development-guide.md:10]: Would it be useful to have a custom Conda `yaml` config for this package?

So a user can just do:

```shell
conda env create -f environment.yml
```

File: docs/sphinx_source/guide/calibration-guide.rst
[the remark this answers is not in the document]
Comment (srivarra) [reply docs/sphinx_source/guide/calibration-guide.rst:1]: I think the `calibration-guide.rst` could be moved into `guide.rst`

**Review by srivarra — CHANGES_REQUESTED** [review]
This is a super great start. Just a few corner cases, and tiny things here and there!

**Comment by amitabhverma** [comment]
> This is a super great start. Just a few corner cases, and tiny things here and there!

Thanks @srivarra for the detailed feedback. This will be super useful for Friday's meeting. I'm not replying to each comment (for now) since we will be discussing this but from my understanding most are related to current limitations of the renderer/formatting, hotlinking video files, etc.

One roadblock I'm facing is the .md links within a .md files do not convert to .html when the html page is being generated causing linking between files an issue unless the .md links are changed to .html. However, in that case they wouldn't work on github. I have tried looking for a solution but could not find something clean.

File: docs/sphinx_source/guide/QLIPP.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# waveorder
```
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/sphinx_source/guide/QLIPP.md]: Yes, we are maintaining single instances. Done.

File: waveorder/cli/apply_inverse_transfer_function.py [hunk 7]
```diff
@@ -9,13 +9,14 @@
 import torch.multiprocessing as mp
 from iohub import open_ome_zarr
 
-from waveorder.cli import apply_inverse_models, jobs_mgmt
```
[the remark this answers is not in the document]
Comment (amitabhverma) [reply waveorder/cli/apply_inverse_transfer_function.py]: Resolved rebasing by Talon, by merging `/main` into the `/docs_update` branch
Confirming `diff` looks clean.

File: docs/sphinx_source/sg_execution_times.rst
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/sphinx_source/sg_execution_times.rst:1]: Not necessary and probably not warranted (in the current form) but was a good example to showcase some possibilities if benchmarking data was to be shared and what benefits can be achieved with parallelization.

File: docs/examples/models/README.rst
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/examples/models/README.rst:1]: The breadcrumbs should be resolving correctly now.
The highlighting issue when clicking under Models & Visuals persists. Needs re-visitation. Clicking on Computation times works as expected so could be a depth issue.

File: docs/sphinx_source/api/stokes.rst
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/sphinx_source/api/stokes.rst:1]: @talonchandler I think we will need your thoughts and inputs on this.

File: docs/examples/models/README.rst
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/examples/models/README.rst:1]: #463

File: docs/sphinx_source/README.md
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/sphinx_source/README.md:1]: #465

File: docs/sphinx_source/api/stokes.rst
[the remark this answers is not in the document]
Comment (talonchandler) [reply docs/sphinx_source/api/stokes.rst:1]: Moved to #466.

File: docs/sphinx_source/README.md (hunk 1 above)
Comment (talonchandler) [reply docs/sphinx_source/README.md]: This `README.md` now only lives in the root of the repo, so I think linking to the github examples makes more sense now.

File: docs/sphinx_source/api/settings.rst
Comment (talonchandler) [reply docs/sphinx_source/api/settings.rst:1]: Long desired and now I can do it! Moved to #468. Thanks @srivarra.

File: docs/sphinx_source/sg_execution_times.rst
Comment (talonchandler) [inline docs/sphinx_source/sg_execution_times.rst:1]: @amitabhverma can you either fix or remove this?

**Review by talonchandler — CHANGES_REQUESTED** [review]
> One roadblock I'm facing is the .md links within a .md files do not convert to .html when the html page is being generated causing linking between files an issue unless the .md links are changed to .html. However, in that case they wouldn't work on github. I have tried looking for a solution but could not find something clean.

Thanks for flagging this @amitabhverma. I think this feature is important, because as is many of the links are broken in the guides. e.g. see the links on this page:
https://waveorder.readthedocs.io/en/latest/guide.html
 
I've implemented a potential fix in https://github.com/mehta-lab/waveorder/pull/469. 

My requests after this round of review are:
- make sure the links work in the guides in both github and the docs website (I think my PR fixes this)
- remove duplicate dependencies, ideally with a single source of truth
- remove or fix the execution times script
- were you able to successfully render a jupyter notebook-style page similar to the iohub examples like we discussed on Friday? No need to make major modifications to the notebook itself...I'm mostly looking for proof of concept. 

Thank you @amitabhverma.

File: docs/sphinx_source/requirements.txt
[the remark this answers is not in the document]
Comment (amitabhverma) [reply docs/sphinx_source/requirements.txt:1]: @talonchandler 
The `requirements.txt` was meant to have only `waveorder` and the docs related dependencies but I see what you mean and agree. I can have a docs section under `project.optional-dependencies` added to `pyproject.toml` and use that and get rid of `requirements.txt`. Would that be ok ?

[the remark this answers is not in the document]
Comment (talonchandler) [reply docs/sphinx_source/requirements.txt:1]: > I can have a docs section under project.optional-dependencies added to pyproject.toml and use that and get rid of requirements.txt. Would that be ok ?

Sounds great to me!

**Review by talonchandler — DISMISSED** [review]

**Review by srivarra — APPROVED** [review]
I think this looks great!
